Mathlib/Data/List/Destutter.lean

```diff
@@ -173,4 +185,187 @@ theorem destutter_eq_nil : ∀ {l : List α}, destutter R l = [] ↔ l = []
+  induction l with
+  | nil => simp
+  | cons c cs ih =>
+    obtain hac | hac : R a c ∨ Rᶜ a c := em _
+    · have hbc : ¬Rᶜ b c := mt (_root_.trans hab) (not_not.2 hac)
+      simp [destutter', if_pos hac, if_pos (not_not.1 hbc)]
+    · have hbc : ¬R b c := trans (symm hab) hac
+      simpa [destutter', if_neg hac, if_neg hbc]
+
+/-- `destutter'` on a relation like ≠, whose negation is an equivalence, has length
+    monotonic under List.cons --/
+theorem le_length_destutter' [IsEquiv α Rᶜ] :
+    (List.destutter' R b l).length ≤ (List.destutter' R a (b :: l)).length := by
+  cases l with
+  | nil => by_cases hab : (R a b) <;> simp_all [Nat.le_succ]
+  | cons c cs =>
+    by_cases hab : (R a b)
+    case pos => simp [destutter', if_pos hab, Nat.le_succ]
+    obtain hac | hac : R a c ∨ Rᶜ a c := em _
+    · have hbc : ¬Rᶜ b c := mt (_root_.trans hab) (not_not.2 hac)
+      simp [destutter', if_pos hac, if_pos (not_not.1 hbc), if_neg hab]
+    · have hbc : ¬R b c := trans (symm hab) hac
+      apply le_of_eq
+      simp only [destutter', if_neg hbc, if_neg hac, if_neg hab]
+      exact (length_destutter' cs R hab).symm
+
+/-- `destutter` on a relation like ≠, whose negation is an equivalence, has length
+    monotonic under List.cons --/
+theorem length_destutter_cons_ge_length_destutter [IsEquiv α Rᶜ] :
+    (l.destutter R).length ≤ ((a::l).destutter R).length := by
+  cases l
+  · simp [destutter]
+  · exact le_length_destutter' _ R
+
+/-- `destutter ≠` has length monotonic under List.cons --/
+theorem length_destutter_ne_cons_ge_length_destutter [DecidableEq α]:
+    (l.destutter (·≠·)).length ≤ ((a::l).destutter (·≠·)).length :=
+  length_destutter_cons_ge_length_destutter l (·≠·)
+
+/-- `destutter'` on a relation like ≠ or <, whose negation is transitive, has length monotonic
+    under a ¬R changing of the first element. -/
+theorem length_destutter'_cotrans_ge [i : IsTrans α Rᶜ] (hba : ¬R b a) :
+    (List.destutter' R b l).length ≤ (List.destutter' R a l).length := by
+  induction l generalizing a with
+  | nil => simp
+  | cons c cs ih =>
+    by_cases hbc : R b c
+    case pos =>
+      have hac : ¬Rᶜ a c := (mt (_root_.trans hba)) (not_not.2 hbc)
+      simp_rw [destutter', if_pos (not_not.1 hac), if_pos hbc, length_cons, le_refl]
+    case neg =>
+      simp only [destutter', if_neg hbc]
+      by_cases hac : R a c
+      case pos =>
+        simp only [if_pos hac, length_cons]
+        exact le_succ_of_le (ih hbc)
+      case neg =>
+        simp only [if_neg hac]
+        exact ih hba
+
+/-- `destutter` of relations like `≠`, whose negation is an equivalence relation,
+    gives a list of maximal length over any chain. In other words: `l.destutter R` is an
+    R-chain sublist of l, and is at least as long as any other R-chain sublist.
+-/
+theorem length_destutter_coequiv_maximal_chain_neg_trans [IsEquiv α Rᶜ]
+    (h₁ : l₂ <+ l) (h₂ : l₂.Chain' R) :
+    l₂.length ≤ (l.destutter R).length := by
```

Comment:
This is a completely unguessable name. Maybe
```suggestion
theorem Chain.length_le_length_destutter [IsEquiv α Rᶜ] (h₁ : l₂ <+ l) (h₂ : l₂.Chain' R) :
    l₂.length ≤ (l.destutter R).length := by
```

Reply:
Thanks. I have just barely enough intuition to guess names sometimes myself. I recognized this name was unguessable, but had no idea how to name it myself. :)